File: parcel-api/parcel-api-go/service.go
package main

import (
	"encoding/json"
	"fmt"
	"math/rand/v2"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// Stores parcel JSON as raw bytes in memory; re-deserialises on every
// request. Mirrors the Rust StubParcelService — putting the JSON parse cost
// on the request path so the bench measures the work a real adapter does
// when decoding data from a downstream.
type service struct {
	parcels [][]byte // raw JSON, sorted by parcel number
	numbers []string // parcel numbers in the same order
}

func loadService(dir string) (*service, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, fmt.Errorf("read parcel data dir %q: %w", dir, err)
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })

	svc := &service{}
	for _, e := range entries {
		name := e.Name()
		if !strings.HasSuffix(name, ".json") {
			continue
		}
		bytes, err := os.ReadFile(filepath.Join(dir, name))
		if err != nil {
			return nil, fmt.Errorf("read %s: %w", name, err)
		}
		stem := strings.TrimSuffix(name, ".json")
		svc.numbers = append(svc.numbers, stem)
		svc.parcels = append(svc.parcels, bytes)
	}
	return svc, nil
}

func (s *service) count() int { return len(s.parcels) }

func (s *service) listParcels() ([]ParcelResponse, error) {
	now := time.Now().UTC()
	order := make([]int, len(s.parcels))
	for i := range order {
		order[i] = i
	}
	rand.Shuffle(len(order), func(i, j int) { order[i], order[j] = order[j], order[i] })

	parcels := make([]ParcelResponse, 0, len(order))
	for _, i := range order {
		var p ParcelResponse
		if err := json.Unmarshal(s.parcels[i], &p); err != nil {
			return nil, fmt.Errorf("parse parcel %s: %w", s.numbers[i], err)
		}
		p.Features = computeFeatures(&p, now)
		parcels = append(parcels, p)
	}
	return parcels, nil
}

File: parcel-api/parcel-api-node/service.js
import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { computeFeatures } from './features.js';
import { serialize, Float } from './serialize.js';

// Stores parcel JSON as raw strings in memory; re-parses on every request.
// Mirrors the other ports — putting the JSON parse + serialize cost on the
// request path so the bench measures the work a real adapter does when decoding
// data from a downstream.
export class StubParcelService {
  constructor(dir) {
    const files = readdirSync(dir)
      .filter((f) => f.endsWith('.json'))
      .sort();
    this.parcelJsons = files.map((f) => readFileSync(join(dir, f), 'utf8'));
  }

  get count() {
    return this.parcelJsons.length;
  }

  listParcels() {
    const nowMs = Date.now();
    const order = this.parcelJsons.map((_, i) => i);
    shuffle(order);

    let out = '[';
    for (let k = 0; k < order.length; k++) {
      if (k > 0) out += ',';
      const parcel = JSON.parse(this.parcelJsons[order[k]]);
      // Features are read from the raw numeric weightInKg, so compute before wrapping.
      parcel.features = computeFeatures(parcel, nowMs);
      if (typeof parcel.weightInKg === 'number') {
        parcel.weightInKg = new Float(parcel.weightInKg);
      }
      out += serialize(parcel);
    }
    return out + ']';
  }
}

function shuffle(arr) {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
}
